lib/util/Components.js

```diff
@@ -670,8 +674,8 @@ function componentRule(rule, context) {
   return updatedRuleInstructions;
 }
 
-Components.detect = function(rule) {
-  return componentRule.bind(this, rule);
+module.exports = {
```

Comment:
this is broken; you're now only exporting an object with `detect`, instead of the entire `Components` constructor.

I think you want `module.exports = Object.assign(Components, { detect(rule) { … } });`

Reply:
I did this on purpose. The way rules interact with this file is:

1/ `const Components = require('../util/Components');`
2/ `create: Components.detect((context, components, utils) => {`

No rule is creating a new instance of the `Components` class directly. Everything related to the `Components` class happens inside `Components.js`. It seems like the API for this file could even just be a single function (e.g. `detectComponents`. Therefore, I don't see the need to export the whole `Components` constructor. Please correct me if I missed anything regarding this file.

The tests are passing and I ran this branch on a fairly large project and didn't have any issues (though I don't have all the rules enabled, so it's not a solid test).

Reply:
`require('eslint-plugin-react/util/Components')` would stop working, however, and that's part of the API.

Reply:
That's correct! 👍 I would have never thought of that... However, what is the use case for this? I don't see any reason why users would use this plugin in such a way.

But yes, if we would make the change that I proposed here, then in theory we'd have to bump the major version. And we don't want to do that for this PR.

The reason I wanted to remove it is because it was quite confusing for me to understand how this module works, and exporting more than what is needed added to the confusion. If possible, it would be good to get rid of it, but for now I'll revert the change so the API doesn't break.

Reply:
I'd say when it's time to make a major bump, this file should just export the detect function directly, if that's what we want.